nexus/src/db/collection_attach.rs

```diff
@@ -0,0 +1,1682 @@
+type SerializedResourcePrimaryKey<ResourceType, C> =
+    <ResourcePrimaryKey<ResourceType, C> as diesel::Expression>::SqlType;
+type SerializedResourceForeignKey<ResourceType, C> =
+    <ResourceForeignKey<ResourceType, C> as diesel::Expression>::SqlType;
+
+type TableSqlType<T> = <T as AsQuery>::SqlType;
+
+type BoxedQuery<T> = diesel::helper_types::IntoBoxed<'static, T, Pg>;
+type BoxedDslOutput<T> = diesel::internal::table_macro::BoxedSelectStatement<
+    'static,
+    TableSqlType<T>,
+    diesel::internal::table_macro::FromClause<T>,
+    Pg,
+>;
+
+/// This implementation uses a CTE which attempts to do the following:
+///
+/// 1. (collection_by_id, resource_by_id): Identify if the collection and
+///    resource objects exist at all.
+/// 2. (resource_count): Identify if the number of resources already attached to
+///    the collection exceeds a threshold.
+/// 3. (collection_info, resource_info): Checks for arbitrary user-provided
+///    constraints on the collection and resource objects.
+/// 4. (do_update): IFF all previous checks succeeded, make a decision to perfom
+///    an update.
+/// 5. (updated_collection):  Increase the generation number on the collection.
+/// 6. (updated_resource): Apply user-provided updates on the resource -
+///    presumably, setting the collection ID value.
+///
+/// This is implemented as follows:
+///
+/// ```text
+/// // WITH
+/// //      /* Look up the collection - Check for existence only! */
+/// //      collection_by_id AS (
+/// //          SELECT * FROM C
+/// //          WHERE <PK> = <VALUE> AND <time_deleted> IS NULL
+/// //          FOR UPDATE
+/// //      ),
+/// //      /* Look up the resource - Check for existence only! */
+/// //      resource_by_id AS (
+/// //          SELECT * FROM R
+/// //          WHERE <PK> = <VALUE> AND <time_deleted> IS NULL
+/// //          FOR UPDATE
+/// //      ),
+/// //      /* Count the number of attached resources */
+/// //      resource_count AS (
+/// //          SELECT COUNT(*) FROM R
+/// //          WHERE <FK> = <VALUE> AND <time_deleted> IS NULL
+/// //      ),
+/// //      /* Look up the collection - Check for additional constraints */
+/// //      collection_info AS (
+/// //          SELECT * FROM C
+/// //          WHERE <PK> = <VALUE> AND <time_deleted> IS NULL AND
+/// //              <Additional user-supplied constraints>
+/// //          FOR UPDATE
+/// //      ),
+/// //      /* Look up the resource - Check for additional constraints */
+/// //      resource_info AS (
+/// //          SELECT * FROM R
+/// //          WHERE <PK> = <VALUE> AND <time_deleted> IS NULL AND
+/// //              <FK> IS NULL AND <Additional user-supplied constraints>
+/// //          FOR UPDATE
+/// //      ),
+/// //      /* Make a decision on whether or not to apply ANY updates */
+/// //      do_update AS (
+/// //          SELECT IF(
+/// //              EXISTS(SELECT id FROM collection_info) AND
```

Comment:
Note: this may be a simpler implementation of the "dueling administrators" problem proposed in RFD 192.

Basically, to perform a conditional UPDATE *and* check for existence of an object, that proposal relies on a division-by-zero to do an early exit before performing the UPDATE itself.

However, it seems possible to:
- Avoid the early exit
- Make the UPDATE conditional on whatever subqueries we want
- Return the results of multiple rows, letting the caller figure out what happened (did the object exist? did we perform the update? etc.)